src/main/host/thread_ptrace.c

```diff
@@ -386,42 +471,164 @@ static pid_t _waitpid_spin(pid_t pid, int *wstatus, int options) {
     return rv;
 }
 
-static void _threadptrace_nextChildState(ThreadPtrace* thread) {
-    // Wait for child to stop.
-
-    int wstatus;
-    if (_waitpid_spin(thread->base.nativeTid, &wstatus, 0) < 0) {
-        error("waitpid: %s", g_strerror(errno));
-        return;
-    }
+// Waits for a ptrace or shim event.
+static StopReason _threadptrace_hybridSpin(ThreadPtrace* thread) {
+    // Only used for an shim-event-stop; lifted out of the loop so we don't
+    // re-initialize on every iteration.
+    StopReason event_stop = {
+        .type = STOPREASON_SHIM_EVENT,
+    };
+    while (1) {
+        if (thread->enableIpc &&
+            shimevent_tryRecvEventFromPlugin(thread->ipcBlk.p, &event_stop.shim_event) == 0) {
```

Comment:
I'm confused here. `shimevent_tryRecvEventFromPlugin` takes a `struct IPCData *data` as its first argument. Why is a pointer to a shared-memory block being used?

Reply:
It compiles because thread->ipcBlk.p is declared type `void *`, and works since it is in this case actually type `struct IPCData*`.

Added a method to get the typed pointer for clarity.